clang/lib/AST/TemplateBase.cpp

```diff
@@ -64,7 +64,19 @@ static void printIntegral(const TemplateArgument &TemplArg, raw_ostream &Out,
         // may create a size difference between the enum value and template
         // argument value, requiring isSameValue here instead of operator==.
         if (llvm::APSInt::isSameValue(ECD->getInitVal(), Val)) {
-          ECD->printQualifiedName(Out, Policy);
+          // Somehow, printQualifiedName does not actually print all the
+          // namespace qualifiers. However, this is needed in the integration
```

Comment:
That kind of seems like a workaround for a particular situation in the integration header, which we suddenly enable for everything. I think `printQualifiedName` intentionally doesn't print all qualifiers for non canonical types because non canonical types are so as written in user's code. Always printing them canonical would be unexpected for default behavior used outside of integration header. IMO, we have to either re-structure how integration header works or modify printing policy so it doesn't break when we set PrintCanonicalTypes there.

Reply:
I made this conditional under a new flag added to Policy.